File: dashboard/reports_api.go
package main

// reports_api.go — the administrator endpoints of the Reports studio (R2):
//
//	GET    /api/reports/templates                  template and element catalog
//	GET    /api/reports/payload-options            #653: payload picker search results
//	GET    /api/reports/definitions                saved definitions (+ ETag)
//	POST   /api/reports/definitions                create (server-assigned id)
//	GET    /api/reports/definitions/{id}           one definition
//	PATCH  /api/reports/definitions/{id}           full-field replace
//	DELETE /api/reports/definitions/{id}
//	POST   /api/reports/definitions/{id}/generate  render now and store the PDF
//	POST   /api/reports/payloads/{hash}/generate   #474: one-click payload PDF, no saved definition
//	GET    /api/reports/generated                  generated history (newest first)
//	GET    /api/reports/generated/{id}/pdf         inline view; ?download=1 attachment
//	DELETE /api/reports/generated/{id}
//
// Every endpoint requires a live-introspected administrator identity through
// the shared guard (same-origin and per-subject write limits on mutations).
// Definition mutations accept an optional If-Match for optimistic
// concurrency, exactly like the settings configuration API. This is the only
// surface that produces PDFs: the per-page export buttons and the legacy
// /export/report.pdf endpoint were removed in R2.

import (
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
	"strings"
	"time"
)

// maxReportBody bounds a definition payload; definitions are small documents.
const maxReportBody = 32 << 10

// reportStudioGuard enforces administrator access and resolves the store.
func (s *store) reportStudioGuard(w http.ResponseWriter, r *http.Request, write bool) bool {
	if s == nil || s.reports == nil {
		http.Error(w, "reports studio unavailable", http.StatusServiceUnavailable)
		return false
	}
	_, ok := s.adminSettingsIdentity(w, r, write)
	return ok
}

func writeReportsJSON(w http.ResponseWriter, etag string, status int, payload any) {
	w.Header().Set("Content-Type", "application/json")
	w.Header().Set("Cache-Control", "no-store")
	if etag != "" {
		w.Header().Set("ETag", etag)
	}
	w.WriteHeader(status)
	_ = json.NewEncoder(w).Encode(payload)
}

// writeReportStoreError maps store failures onto the settings API contract's
// status codes (conflicts, validation, missing records, read-only stores stay
// distinguishable without parsing text) -- reports_store.go returns the same
// sentinel errors settings_store.go does. Not a passthrough to
// writePreferenceError: that hardcoded "preferences changed concurrently;
// reload and retry" for errStaleRevision, which is the wrong subsystem's name
// on a report-definition conflict (#211) and reads as evidence of a bug in
// the wrong place entirely.
func writeReportStoreError(w http.ResponseWriter, err error) {
	switch {
	case errors.Is(err, errStaleRevision):
		http.Error(w, "this report definition changed concurrently; reload and retry", http.StatusConflict)
	case errors.Is(err, errSettingsValidation):
		// Validation messages name fields and bounds, never values.
		http.Error(w, err.Error(), http.StatusUnprocessableEntity)
	case errors.Is(err, errUnknownRecord):
		http.Error(w, "no reports record yet; reload the page first", http.StatusNotFound)
	case errors.Is(err, errStoreReadOnly):
		http.Error(w, "reports store is read-only; contact an administrator", http.StatusServiceUnavailable)
	case errors.Is(err, errReportsStorageUnavailable):
		http.Error(w, "generated-report storage is unavailable; Elasticsearch is not configured", http.StatusServiceUnavailable)
	default:
		http.Error(w, "report definition could not be saved", http.StatusInternalServerError)
	}
}

func (s *store) serveReportTemplates(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodGet {
		w.Header().Set("Allow", "GET")
		http.Error(w, "GET required", http.StatusMethodNotAllowed)
		return
	}
	if !s.reportStudioGuard(w, r, false) {
		return
	}
	var overrides map[string]reportPresetOverride
	if s.settings != nil {
		cfg, _ := s.settings.config.Get()
		overrides = cfg.Presentation.ReportPresets
	}
	writeReportsJSON(w, "", http.StatusOK, map[string]any{
		"templates": reportTemplateCatalogWithOverrides(overrides),
		"elements":  reportElementCatalog,
		"windows":   []string{"1h", "6h", "24h", "7d", "30d"},
		"themes":    []string{"dark", "light"},
	})
}

// reportPayloadOptionsLimit bounds /api/reports/payload-options the same way
// every other bounded list on this dashboard is capped (25-30 rows) -- a
// picker narrows a search, it does not browse the full inventory.
const reportPayloadOptionsLimit = 30

// reportPayloadOption is one row of Report Studio's payload picker (#653):
// enough to identify a captured payload plus, per the issue's explicit ask,
// which analysis sources already exist for it (correlateHash, the same
// cross-source check payload-workbench already uses), so an operator can
// tell at a glance which payloads have enough analysis depth to make a
// meaningful report before picking one.
type reportPayloadOption struct {
	Hash     string `json:"hash"`
	Kind     string `json:"kind"`
	SizeH    string `json:"size_h"`
	MtimeUTC string `json:"mtime_utc"`
	Sandbox  bool   `json:"sandbox"`
	Ghidra   bool   `json:"ghidra"`
	GitHub   bool   `json:"github"`
}

// serveReportPayloadOptions backs the payload template's picker in the
// designer (#653) -- the payload template requires scope.hash
// (reports_store.go's validateDefinition), and until this endpoint existed
// there was no way to search captured payloads and see what to set it to
// from inside Report Studio at all; the only working path was the separate
// /api/reports/payloads/{hash}/generate shortcut reached from an individual
// payload's own analysis page, which bypasses the designer entirely.
func (s *store) serveReportPayloadOptions(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodGet {
		w.Header().Set("Allow", "GET")
		http.Error(w, "GET required", http.StatusMethodNotAllowed)
		return
	}
	if !s.reportStudioGuard(w, r, false) {
		return
	}
	// payloadsData, not a direct disk scan: this dashboard reads payload
	// inventory through Elasticsearch only (the same rule #403 established
	// for sensor/event data) -- payload-inventory-worker owns the disk walk
	// itself (#1201/#1223), refreshPayloadCacheAsync just reads its output
	// back, never called directly from a request handler.
	data := s.payloadsData(payloadsFilter{Hash: r.URL.Query().Get("q")})
	files := data.Files
	if len(files) > reportPayloadOptionsLimit {
		files = files[:reportPayloadOptionsLimit]
	}
	// Loaded once for the whole search, not once per file below: see
	// correlateHash's own doc comment -- each of these fetches up to 10000
	// documents from ES regardless of hash, so calling it inside this loop
	// (as this used to) meant up to reportPayloadOptionsLimit separate
	// fetches per source for one search-as-you-type keystroke.
	ghidraResults, sandboxResults, githubResults := loadGhidraResults(), loadSandboxResults(), loadGitHubAnalysisResults()
	options := make([]reportPayloadOption, 0, len(files))
	for _, file := range files {
		correlation := s.correlateHash(file.Hash, "", ghidraResults, sandboxResults, githubResults)
		options = append(options, reportPayloadOption{
			Hash:     file.Hash,
			Kind:     file.Kind,
			SizeH:    file.SizeH,
			MtimeUTC: file.MtimeUTC,
			Sandbox:  len(correlation.Sandbox) > 0,
			Ghidra:   correlation.Ghidra != nil,
			GitHub:   correlation.GitHub != nil,
		})
	}
	writeReportsJSON(w, "", http.StatusOK, map[string]any{"payloads": options})
}

func (s *store) serveReportDefinitions(w http.ResponseWriter, r *http.Request) {
	switch r.Method {
	case http.MethodGet:
		if !s.reportStudioGuard(w, r, false) {
			return
		}
		doc, etag := s.reports.document()
		writeReportsJSON(w, etag, http.StatusOK, map[string]any{"definitions": doc.Definitions})
	case http.MethodPost:
		s.createReportDefinition(w, r)
	default:
		w.Header().Set("Allow", "GET, POST")
		http.Error(w, "GET or POST required", http.StatusMethodNotAllowed)
	}
}

func (s *store) createReportDefinition(w http.ResponseWriter, r *http.Request) {
	if !s.reportStudioGuard(w, r, true) {
		return
	}
	def, ok := decodeReportDefinition(w, r)
	if !ok {
		return
	}
	if def.ID != "" {
		http.Error(w, "id is assigned by the server; omit it when creating a definition", http.StatusBadRequest)
		return
	}
	created, etag, err := s.reports.putDefinition(r.Header.Get("If-Match"), def)
	if err != nil {
		writeReportStoreError(w, err)
		return
	}
	writeReportsJSON(w, etag, http.StatusCreated, map[string]any{"definition": created})
}

func decodeReportDefinition(w http.ResponseWriter, r *http.Request) (reportDefinition, bool) {
	if contentType := strings.ToLower(strings.TrimSpace(r.Header.Get("Content-Type"))); !strings.HasPrefix(contentType, "application/json") {
		http.Error(w, "Content-Type: application/json required", http.StatusUnsupportedMediaType)
		return reportDefinition{}, false
	}
	body, err := io.ReadAll(http.MaxBytesReader(w, r.Body, maxReportBody))
	if err != nil {
		http.Error(w, "request body too large or unreadable", http.StatusRequestEntityTooLarge)
		return reportDefinition{}, false
	}
	var def reportDefinition
	if !strictDecode(body, &def) {
		http.Error(w, "invalid or unknown report definition fields", http.StatusBadRequest)
		return reportDefinition{}, false
	}
	return def, true
}

func (s *store) serveReportDefinitionByID(w http.ResponseWriter, r *http.Request) {
	tail := strings.TrimPrefix(r.URL.Path, "/api/reports/definitions/")
	id, action, _ := strings.Cut(tail, "/")
	if !validReportID(id, "rep_") {
		http.NotFound(w, r)
		return
	}
	switch action {
	case "":
		switch r.Method {
		case http.MethodGet:
			if !s.reportStudioGuard(w, r, false) {
				return
			}
			def, ok := s.reports.definition(id)
			if !ok {
				http.NotFound(w, r)
				return
			}
			writeReportsJSON(w, "", http.StatusOK, map[string]any{"definition": def})
		case http.MethodPatch:
			s.replaceReportDefinition(w, r, id)
		case http.MethodDelete:
			s.removeReportDefinition(w, r, id)
		default:
			w.Header().Set("Allow", "GET, PATCH, DELETE")
			http.Error(w, "GET, PATCH, or DELETE required", http.StatusMethodNotAllowed)
		}
	case "generate":
		if r.Method != http.MethodPost {
			w.Header().Set("Allow", "POST")
			http.Error(w, "POST required", http.StatusMethodNotAllowed)
			return
		}
		s.generateReport(w, r, id, "manual")
	default:
		http.NotFound(w, r)
	}
}

func (s *store) replaceReportDefinition(w http.ResponseWriter, r *http.Request, id string) {
	if !s.reportStudioGuard(w, r, true) {
		return
	}
	def, ok := decodeReportDefinition(w, r)
	if !ok {
		return
	}
	if def.ID != "" && def.ID != id {
		http.Error(w, "id in the body must match the path or be omitted", http.StatusBadRequest)
		return
	}
	def.ID = id
	updated, etag, err := s.reports.putDefinition(r.Header.Get("If-Match"), def)
	if err != nil {
		writeReportStoreError(w, err)
		return
	}
	writeReportsJSON(w, etag, http.StatusOK, map[string]any{"definition": updated})
}

func (s *store) removeReportDefinition(w http.ResponseWriter, r *http.Request, id string) {
	if !s.reportStudioGuard(w, r, true) {
		return
	}
	if err := s.reports.deleteDefinition(r.Header.Get("If-Match"), id); err != nil {
		writeReportStoreError(w, err)
		return
	}
	_, etag := s.reports.document()
	writeReportsJSON(w, etag, http.StatusOK, map[string]any{"deleted": id})
}

// generateReport renders a definition immediately and records the PDF. The
// scheduled path (R4) shares this renderer with origin "schedule".
func (s *store) generateReport(w http.ResponseWriter, r *http.Request, id, origin string) {
	if !s.reportStudioGuard(w, r, true) {
		return
	}
	meta, etag, err := s.renderDefinitionToStored(id, origin)
	if err != nil {
		writeReportStoreError(w, err)
		return
	}
	writeReportsJSON(w, etag, http.StatusCreated, map[string]any{"generated": meta})
}

// serveGeneratePayloadReport backs #474's payload-page "Generate PDF"
// button: POST /api/reports/payloads/{hash}/generate. Synchronous, like
// generateReport above -- report_pdf.go's writer is an in-process, hand-
// rolled PDF emitter (no external process to poll), so there is no
// background job to report progress on; the caller's own fetch await is the
// only "progress indicator" needed.
func (s *store) serveGeneratePayloadReport(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodPost {
		w.Header().Set("Allow", http.MethodPost)
		http.Error(w, "POST required", http.StatusMethodNotAllowed)
		return
	}
	if !s.reportStudioGuard(w, r, true) {
		return
	}
	hash := strings.TrimSuffix(strings.TrimPrefix(r.URL.Path, "/api/reports/payloads/"), "/generate")
	meta, etag, err := s.generatePayloadReport(hash, "manual")
	if err != nil {
		writeReportStoreError(w, err)
		return
	}
	writeReportsJSON(w, etag, http.StatusCreated, map[string]any{"generated": meta})
}

// renderDefinitionToStored is the generation pipeline shared by the manual
// API and the scheduler: resolve the definition, render the PDF from live
// telemetry (or the referenced sandbox/payload artifact), and persist it.
func (s *store) renderDefinitionToStored(id, origin string) (generatedReport, string, error) {
	// #787: definitionFresh, not definition -- generating right after saving
	// a definition is the normal designer flow, and the two requests can
	// land on different dashboard replicas. See GetFresh's own comment.
	def, ok := s.reports.definitionFresh(id)
	if !ok {
		return generatedReport{}, "", fmt.Errorf("%w: no report definition with this id", errUnknownRecord)
	}
	pdf, title, err := s.renderDefinitionPDFBytes(def)
	if err != nil {
		return generatedReport{}, "", err
	}
	return s.reports.addGenerated(generatedReport{
		DefinitionID: def.ID,
		Name:         def.Name,
		Template:     def.Template,
		Theme:        def.Theme,
		Title:        title,
		Origin:       origin,
	}, pdf)
}

// renderDefinitionPDFBytes is renderDefinitionToStored's rendering step,
// pulled out on its own so #474's payload report can share it against an
// ephemeral, never-persisted definition (generatePayloadReport below)
// instead of requiring a saved definition id like the designer flow does.
func (s *store) renderDefinitionPDFBytes(def reportDefinition) ([]byte, string, error) {
	template, _ := reportTemplateByID(def.Template)
	title := strings.TrimSpace(def.Branding.Title)
	if title == "" {
		title = template.Title
	}
	now := time.Now()
	switch {
	case template.Sandbox:
		data, err := sandboxData(def.Scope.Job, "")
		if err != nil || data.Detail == nil {
			return nil, "", fmt.Errorf("%w: scope.job does not resolve to a completed sandbox result", errSettingsValidation)
		}
		return renderThemedSandboxReportPDF(*data.Detail, now, pdfThemeNamed(def.Theme), def.Branding.pdf()), title, nil
	case template.Payload:
		analysis, err := s.analyzePayload(def.Scope.Hash)
		if err != nil {
			return nil, "", fmt.Errorf("%w: scope.hash does not resolve to a captured payload", errSettingsValidation)
		}
		return renderThemedPayloadReportPDF(analysis, now, pdfThemeNamed(def.Theme), def.Branding.pdf()), title, nil
	case template.Ghidra:
		data, err := s.ghidraData(def.Scope.Hash, "")
		if err != nil || data.Detail == nil {
			return nil, "", fmt.Errorf("%w: scope.hash does not resolve to a completed ghidra result", errSettingsValidation)
		}
		return renderThemedGhidraReportPDF(*data.Detail, now, pdfThemeNamed(def.Theme), def.Branding.pdf()), title, nil
	default:
		data := s.reportDataFor(def.Scope.filter(now))
		data.Title = title
		return renderDefinitionPDF(data, def), title, nil
	}
}

// generatePayloadReport is #474's one-click "Generate PDF" trigger on the
// payload detail page: unlike the designer flow, it never requires the
// operator to first create and save a named report definition. It builds an
// ephemeral, never-persisted definition scoped to this one hash and reuses
// the exact same rendering/storage/serving pipeline every other Reports
// studio PDF goes through -- the generated record and its file land in the
// same reports document and are viewable/deletable from Reports studio like
// any other, they just have no DefinitionID (nothing to look back up).
func (s *store) generatePayloadReport(hash, origin string) (generatedReport, string, error) {
	if !hashName.MatchString(hash) {
		return generatedReport{}, "", fmt.Errorf("%w: invalid payload id", errSettingsValidation)
	}
	def := reportDefinition{Template: "payload", Theme: "dark", Scope: reportScope{Hash: hash}}
	pdf, title, err := s.renderDefinitionPDFBytes(def)
	if err != nil {
		return generatedReport{}, "", err
	}
	return s.reports.addGenerated(generatedReport{
		Name:     "Payload " + hash,
		Template: def.Template,
		Theme:    def.Theme,
		Title:    title,
		Origin:   origin,
	}, pdf)
}

func (s *store) serveReportsGenerated(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodGet {
		w.Header().Set("Allow", "GET")
		http.Error(w, "GET required", http.StatusMethodNotAllowed)
		return
	}
	if !s.reportStudioGuard(w, r, false) {
		return
	}
	all, err := s.reports.listGenerated()
	if err != nil {
		writeReportStoreError(w, err)
		return
	}
	// listGenerated returns oldest first; the API serves newest first.
	ordered := make([]generatedReport, len(all))
	for index, meta := range all {
		ordered[len(all)-1-index] = meta
	}
	writeReportsJSON(w, "", http.StatusOK, map[string]any{"generated": ordered})
}

func (s *store) serveReportGeneratedByID(w http.ResponseWriter, r *http.Request) {
	tail := strings.TrimPrefix(r.URL.Path, "/api/reports/generated/")
	id, action, _ := strings.Cut(tail, "/")
	if !validReportID(id, "gen_") {
		http.NotFound(w, r)
		return
	}
	switch action {
	case "pdf":
		if r.Method != http.MethodGet {
			w.Header().Set("Allow", "GET")
			http.Error(w, "GET required", http.StatusMethodNotAllowed)
			return
		}
		s.serveGeneratedPDF(w, r, id)
	case "":
		if r.Method != http.MethodDelete {
			w.Header().Set("Allow", "DELETE")
			http.Error(w, "DELETE required", http.StatusMethodNotAllowed)
			return
		}
		if !s.reportStudioGuard(w, r, true) {
			return
		}
		if err := s.reports.deleteGenerated(id); err != nil {
			writeReportStoreError(w, err)
			return
		}
		writeReportsJSON(w, "", http.StatusOK, map[string]any{"deleted": id})
	default:
		http.NotFound(w, r)
	}
}

func (s *store) serveGeneratedPDF(w http.ResponseWriter, r *http.Request, id string) {
	if !s.reportStudioGuard(w, r, false) {
		return
	}
	meta, body, err := s.reports.generatedPDF(id)
	if err != nil {
		if errors.Is(err, errUnknownRecord) {
			http.NotFound(w, r)
			return
		}
		http.Error(w, "generated report is no longer available", http.StatusGone)
		return
	}
	disposition := "inline"
	if r.URL.Query().Get("download") == "1" {
		disposition = "attachment"
	}
	w.Header().Set("Content-Type", "application/pdf")
	w.Header().Set("Content-Disposition", disposition+`; filename="`+reportDownloadName(meta)+`"`)
	w.Header().Set("Cache-Control", "no-store")
	w.Header().Set("X-Content-Type-Options", "nosniff")
	w.Write(body)
}

// reportDownloadName derives a filesystem-safe download name from the report
// metadata: the design name plus the generation timestamp.
func reportDownloadName(meta generatedReport) string {
	name := strings.ToLower(meta.Name)
	if name == "" {
		name = meta.Template
	}
	var clean strings.Builder
	dash := false
	for _, char := range name {
		if char >= 'a' && char <= 'z' || char >= '0' && char <= '9' {
			clean.WriteRune(char)
			dash = false
		} else if !dash {
			clean.WriteByte('-')
			dash = true
		}
	}
	base := strings.Trim(clean.String(), "-")
	if base == "" {
		base = "report"
	}
	return "honeypot-" + base + "-" + meta.CreatedAt.UTC().Format("20060102-1504") + ".pdf"
}

File: arcane/home/honeypot-dashboard/backend-service/src/reports_api.rs
//! Reports studio HTTP layer (#1612): template/element catalog, definitions
//! CRUD, and the on-demand generate trigger. Ported from
//! reports_api.go's serveReportTemplates/serveReportDefinitions/
//! createReportDefinition/serveReportDefinitionByID/replaceReportDefinition/
//! removeReportDefinition/generateReport/renderDefinitionToStored/
//! renderDefinitionPDFBytes — only the `default` (generic telemetry)
//! branch of renderDefinitionPDFBytes; the sandbox/payload/ghidra
//! artifact-referenced templates are validated but not yet rendered here
//! (see reports_store.rs's module doc comment for the scope decision).
//!
//! No admin-role/same-origin/If-Match machinery — this crate's trust
//! boundary is the BFF's service token, same posture as every other write
//! path here (config.rs/preferences.rs).

use axum::{
    extract::{Path, State},
    http::StatusCode,
    Json,
};
use serde::Deserialize;
use serde_json::{json, Value};

use crate::reports_store::{
    self, put_definition, report_template_catalog, GeneratedReportMeta, ReportDefinition,
    REPORT_ELEMENT_CATALOG,
};
use crate::AppState;

fn bad_request(message: impl Into<String>) -> (StatusCode, String) {
    (StatusCode::BAD_REQUEST, message.into())
}

fn bad_gateway(error: anyhow::Error) -> (StatusCode, String) {
    (StatusCode::BAD_GATEWAY, error.to_string())
}

pub async fn templates() -> Json<Value> {
    let templates: Vec<Value> = report_template_catalog()
        .into_iter()
        .map(|t| {
            json!({
                "id": t.id, "name": t.name, "description": t.description, "title": t.title,
                "theme": t.theme, "window": t.window, "elements": t.elements,
                "sandbox": t.sandbox, "payload": t.payload, "ghidra": t.ghidra,
            })
        })
        .collect();
    let elements: Vec<Value> = REPORT_ELEMENT_CATALOG
        .iter()
        .map(|e| json!({"id": e.id, "label": e.label, "description": e.description}))
        .collect();
    Json(json!({"templates": templates, "elements": elements}))
}

pub async fn list_definitions(
    State(state): State<AppState>,
) -> Result<Json<Value>, (StatusCode, String)> {
    let definitions = reports_store::list_definitions(&state)
        .await
        .map_err(bad_gateway)?;
    Ok(Json(json!({"definitions": definitions})))
}

pub async fn get_definition(
    State(state): State<AppState>,
    Path(id): Path<String>,
) -> Result<Json<Value>, (StatusCode, String)> {
    let definition = reports_store::get_definition(&state, &id)
        .await
        .map_err(bad_gateway)?
        .ok_or((
            StatusCode::NOT_FOUND,
            "no such report definition".to_string(),
        ))?;
    Ok(Json(json!({"definition": definition})))
}

pub async fn create_definition(
    State(state): State<AppState>,
    Json(mut def): Json<ReportDefinition>,
) -> Result<(StatusCode, Json<Value>), (StatusCode, String)> {
    if !def.id.is_empty() {
        return Err(bad_request(
            "id is assigned by the server; omit it when creating a definition",
        ));
    }
    def.id = String::new();
    let created = put_definition(&state, def).await.map_err(map_store_error)?;
    Ok((StatusCode::CREATED, Json(json!({"definition": created}))))
}

pub async fn replace_definition(
    State(state): State<AppState>,
    Path(id): Path<String>,
    Json(mut def): Json<ReportDefinition>,
) -> Result<Json<Value>, (StatusCode, String)> {
    if !def.id.is_empty() && def.id != id {
        return Err(bad_request(
            "id in the body must match the path or be omitted",
        ));
    }
    def.id = id;
    let updated = put_definition(&state, def).await.map_err(map_store_error)?;
    Ok(Json(json!({"definition": updated})))
}

pub async fn delete_definition(
    State(state): State<AppState>,
    Path(id): Path<String>,
) -> Result<Json<Value>, (StatusCode, String)> {
    reports_store::delete_definition(&state, &id)
        .await
        .map_err(map_store_error)?;
    Ok(Json(json!({"deleted": id})))
}

fn map_store_error(message: String) -> (StatusCode, String) {
    if message.contains("no report definition with this id") {
        (StatusCode::NOT_FOUND, message)
    } else if message.contains("limit reached") {
        (StatusCode::CONFLICT, message)
    } else {
        (StatusCode::UNPROCESSABLE_ENTITY, message)
    }
}

#[derive(Deserialize)]
pub struct GenerateBody {
    #[serde(default = "default_origin")]
    origin: String,
}

fn default_origin() -> String {
    "manual".into()
}

pub async fn generate(
    State(state): State<AppState>,
    Path(id): Path<String>,
    body: Option<Json<GenerateBody>>,
) -> Result<(StatusCode, Json<Value>), (StatusCode, String)> {
    let origin = body
        .map(|b| b.origin.clone())
        .unwrap_or_else(default_origin);
    let definition = reports_store::get_definition(&state, &id)
        .await
        .map_err(bad_gateway)?
        .ok_or((
            StatusCode::NOT_FOUND,
            "no such report definition".to_string(),
        ))?;
    let meta = render_definition_to_stored(&state, &definition, &origin)
        .await
        .map_err(map_render_error)?;
    Ok((StatusCode::CREATED, Json(json!({"generated": meta}))))
}

fn map_render_error(message: String) -> (StatusCode, String) {
    if message.contains("not yet implemented") {
        (StatusCode::NOT_IMPLEMENTED, message)
    } else if message.contains("does not resolve") {
        (StatusCode::UNPROCESSABLE_ENTITY, message)
    } else {
        (StatusCode::BAD_GATEWAY, message)
    }
}

/// Up to 20 sandbox-analysis-v1 runs matching a captured payload's hash,
/// newest first — payload_pdf.go's own `binaryAnalysis.SandboxRuns` slice,
/// assembled here from the promoted `file.hash.sha256` field every
/// es_importer.rs-mirrored source carries (same field ghidra_run/cape_run
/// query elsewhere in this crate). Raw `_source` docs, same shape
/// render_sandbox_pdf itself reads.
async fn sandbox_runs_for_hash(state: &AppState, hash: &str) -> anyhow::Result<Vec<Value>> {
    let result = state
        .es
        .search_index(
            &["sandbox-analysis-v1"],
            json!({"size": 20, "sort": [{"@timestamp": {"order": "desc"}}], "query": {"term": {"file.hash.sha256": hash}}}),
        )
        .await?;
    Ok(result["hits"]["hits"].as_array().into_iter().flatten().map(|hit| hit["_source"].clone()).collect())
}

/// renderDefinitionToStored + renderDefinitionPDFBytes, shared by the
/// manual generate endpoint and the scheduler (worker.rs). The sandbox/
/// payload/ghidra branches each resolve their scope reference to one
/// artifact and hand it to report_pdf.rs's matching renderer; `default`
/// keeps the generic telemetry path unchanged.
pub async fn render_definition_to_stored(
    state: &AppState,
    def: &ReportDefinition,
    origin: &str,
) -> Result<GeneratedReportMeta, String> {
    let template = crate::reports_store::report_template_catalog()
        .into_iter()
        .find(|t| t.id == def.template)
        .ok_or_else(|| "unknown report template".to_string())?;

    let title = {
        let trimmed = def.branding.title.trim();
        if trimmed.is_empty() {
            template.title.to_string()
        } else {
            trimmed.to_string()
        }
    };
    let theme = crate::report_pdf::pdf_theme_named(&def.theme);
    let branding = def.branding.to_pdf_branding();
    let generated = chrono::Utc::now();

    let pdf = if template.sandbox {
        let job = def.scope.job.trim();
        if job.is_empty() {
            return Err("scope.job does not resolve to a completed sandbox result".to_string());
        }
        let doc = crate::detail::sandbox_run(State(state.clone()), Path(job.to_string()))
            .await
            .map_err(|_| "scope.job does not resolve to a completed sandbox result".to_string())?
            .0;
        crate::report_pdf::render_sandbox_pdf(&doc, generated, theme, branding)
    } else if template.ghidra {
        let hash = def.scope.hash.trim();
        if hash.is_empty() {
            return Err("scope.hash does not resolve to a completed ghidra result".to_string());
        }
        let doc = crate::detail::ghidra_run(State(state.clone()), Path(hash.to_string()))
            .await
            .map_err(|_| "scope.hash does not resolve to a completed ghidra result".to_string())?
            .0;
        crate::report_pdf::render_ghidra_pdf(&doc, generated, theme, branding)
    } else if template.payload {
        let hash = def.scope.hash.trim();
        if hash.is_empty() {
            return Err("scope.hash does not resolve to a captured payload".to_string());
        }
        let detail = crate::payload_detail::detail(State(state.clone()), Path(hash.to_string()))
            .await
            .map_err(|_| "scope.hash does not resolve to a captured payload".to_string())?
            .0;
        let inventory = detail.inventory.unwrap_or(Value::Null);
        let analysis = detail.analysis.unwrap_or(Value::Null);
        let sandbox_runs = sandbox_runs_for_hash(state, hash).await.map_err(|e| e.to_string())?;
        let github_analysis = crate::detail::github_analysis_run(State(state.clone()), Path(hash.to_string()))
            .await
            .map(|json| json.0)
            .unwrap_or(Value::Null);
        crate::report_pdf::render_payload_pdf(
            hash,
            &inventory,
            &analysis,
            &detail.yara,
            &sandbox_runs,
            &github_analysis,
            generated,
            theme,
            branding,
        )
    } else {
        let appendix_limit = if def.appendix_limit <= 0 { 120 } else { def.appendix_limit };
        let mut data = crate::reports_data::report_data_for(state, &def.scope, title.clone(), appendix_limit)
            .await
            .map_err(|e| e.to_string())?;
        data.title = title.clone();
        crate::report_pdf::render_report_pdf(&data, theme, branding, &def.elements, appendix_limit)
    };

    reports_store::add_generated(
        state,
        GeneratedReportMeta {
            id: String::new(),
            definition_id: def.id.clone(),
            name: def.name.clone(),
            template: def.template.clone(),
            theme: def.theme.clone(),
            title,
            size_bytes: 0,
            created_at: String::new(),
            origin: origin.to_string(),
        },
        pdf,
    )
    .await
    .map_err(|e| e.to_string())
}
